### PR TITLE
Proposal of adding a daemon to send a mail of done queries

### PR DESCRIPTION
**Pull request by Mathoxs**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -3,7 +3,7 @@
     <parent>
         <artifactId>lutece-global-pom</artifactId>
         <groupId>fr.paris.lutece.tools</groupId>
-        <version>3.0.11</version>
+        <version>4.0.3</version>
```
Comment (rzara) [inline pom.xml]: This version is one year old. Why not 4.0.5 ?

File: src/java/fr/paris/lutece/plugins/searchstats/business/IQueryRecordDAO.java [hunk 2]
```diff
@@ -1,4 +1,4 @@
-/*
+	/*
```
Comment (rzara) [inline src/java/fr/paris/lutece/plugins/searchstats/business/IQueryRecordDAO.java:1]: why this change ?

File: src/java/fr/paris/lutece/plugins/searchstats/business/IQueryRecordDAO.java [hunk 3]
```diff
@@ -1,4 +1,4 @@
-/*
+	/*
  * Copyright (c) 2002-2014, Mairie de Paris
```
Comment (rzara) [inline src/java/fr/paris/lutece/plugins/searchstats/business/IQueryRecordDAO.java:2]: Maybe update the year there

File: webapp/WEB-INF/plugins/searchstats.xml [hunk 4]
```diff
@@ -33,5 +33,13 @@
             <dashboard-zone-order>2</dashboard-zone-order>
         </dashboard-component>
     </dashboard-components>
+    <daemons>
+        <daemon>
+            <daemon-id>StatsDaemon</daemon-id>
```
Comment (rzara) [inline webapp/WEB-INF/plugins/searchstats.xml]: This seems like an overly generic name for a daemon, especially for a generic plugin. I guess there are many contexts where a StatsDaemon makes sense. Maybe rename to SearchStatsDaemon ?

File: webapp/WEB-INF/templates/admin/plugins/searchstats/mail/mail_content.html [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+<table>
+	<tr>
+		<th>#i18n{${i18nKeyType}}</th>
+		<th>#i18n{${i18nKeyOccurence}}</th>
+	</tr>
+	<#list listResultRecord as resultRecord>
+	<#if (resultRecord.term)?? && (resultRecord?counter lt nbTopResultMax) >
+		<tr>
+			<td>${(resultRecord.term)!''}</td>
```
Comment (rzara) [inline webapp/WEB-INF/templates/admin/plugins/searchstats/mail/mail_content.html]: You should html escape the `resultRecord.term`.
Besides, you just checked that `resultRecord.term` has content, so `!''` seems unnecessary.

File: src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+
+public final class SearchStatsService
+{
+
+    private static final String PROPERTY_MAIL_SENDER_NAME = AppPropertiesService.getProperty( "daemon.mail.sender_name" );
```
Comment (rzara) [inline src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java]: You should not turn this property into a constant. this prevents updating the value while the app is running

File: src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    private static final String CONSTANT_MAIL_SENDER = MailService.getNoReplyEmail( );
+    private static final String MESSAGE_MAIL_SUBJECT = I18nService.getLocalizedString("searchstats.mail_content.subject", Locale.getDefault());
```
Comment (rzara) [inline src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java]: Same here. I doubt there is any perf concern here

File: src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+
+
+public final class SearchStatsService
+{
+
+    private static final String PROPERTY_MAIL_SENDER_NAME = AppPropertiesService.getProperty( "daemon.mail.sender_name" );
+    private static final String CONSTANT_MAIL_SENDER = MailService.getNoReplyEmail( );
```
Comment (rzara) [inline src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java]: same here

File: src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        for ( Recipient recipient : colMailingList )
+        {
+            MailService.sendMailHtml( recipient.getEmail( ), PROPERTY_MAIL_SENDER_NAME, CONSTANT_MAIL_SENDER, MESSAGE_MAIL_SUBJECT, strContent );
+        }
+
+    }
+
+    /**
+     * @param nYear
+     *            nMonth nDay nHour The limit Year / Month / Day / Hour
```
Comment (rzara) [inline src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java]: this is invalid javadoc comment

File: src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+     * @param listQueryRecords
+     *            The list of Query Records
+     * @param nLimit
+     *            The limit number of queries
+     * @return The list of the QueryRecords
+     */
+    public List<TermRecord> getMapSearchStatsAfterdate( int nYear, int nMonth, int nDay, int nHour, List<QueryRecord> listQueryRecords, int nLimit )
```
Comment (rzara) [inline src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java]: What does `Map` mean in this method name ?

File: src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+    /**
+     * @return The timeRecord
+     */
+    public static TimedRecord getIntervalTimeRecord( )
+    {
+        TimedRecord timedRecord = new TimedRecord( );
+        String strNombreSecondes = DatastoreService.getInstanceDataValue( STATS_DAEMON_INTERVAL, DEFAULT_INTERVAL );
+        int nStrNombreSecondes = Integer.parseInt( strNombreSecondes );
+        // Converti les secondes en Y/M/D/H
+        int nYear = ( nStrNombreSecondes / 31536000 );
```
Comment (rzara) [inline src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java]: What's this magic number ?

File: src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    /**
+     * @return The timeRecord
+     */
+    public static TimedRecord getIntervalTimeRecord( )
+    {
+        TimedRecord timedRecord = new TimedRecord( );
+        String strNombreSecondes = DatastoreService.getInstanceDataValue( STATS_DAEMON_INTERVAL, DEFAULT_INTERVAL );
+        int nStrNombreSecondes = Integer.parseInt( strNombreSecondes );
```
Comment (rzara) [inline src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java]: you should drop the `Str` from the variable name

File: src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+        // Converti les secondes en Y/M/D/H
+        int nYear = ( nStrNombreSecondes / 31536000 );
+        int nMonth = ( ( nStrNombreSecondes - ( nYear * 31536000 ) ) / 2628000 );
+        int nDay = ( ( nStrNombreSecondes - ( nYear * 31536000 + nMonth * 2628000 ) ) / 86400 );
+        int nHour = ( ( nStrNombreSecondes - ( nYear * 31536000 + nMonth * 2628000 + nDay * 86400 ) ) / 3600 );
```
Comment (rzara) [inline src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java]: same for all magic numbers

File: src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+            timeReport.setDay( 1 );
+        }
+        return getMapSearchStatsAfterdate( timeReport.getYear( ), timeReport.getMonth( ), timeReport.getDay( ), timeReport.getHour( ), listQueryRecords, nLimit );
+    }
+
+    /**
+     * @param timeReport
+     *            The time from which the query can be taken into account
+     * @param query
+     * @return Verify if the queriy has been done after the timeReport
```
Comment (rzara) [inline src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java]: typo

File: src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+     */
+
+    public static boolean checkAfterDate( TimedRecord timeReport, QueryRecord query )
+
+    {
+        LocalDateTime compare = LocalDateTime.now( ).minus( timeReport.getYear( ), ChronoUnit.YEARS ).minus( timeReport.getMonth( ), ChronoUnit.MONTHS )
+                .minus( timeReport.getDay( ), ChronoUnit.DAYS ).minus( timeReport.getHour( ), ChronoUnit.HOURS );
+        LocalDateTime queryLocalTime = LocalDateTime.of( query.getYear( ), query.getMonth( ), query.getDay( ), query.getHour( ), 0 );
+        boolean rep = ( queryLocalTime.compareTo( compare ) >= 0 );
```
Comment (rzara) [inline src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java]: why the variable ?

File: src/java/fr/paris/lutece/plugins/searchstats/daemon/StatsDaemon.java [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+import fr.paris.lutece.plugins.searchstats.service.SearchStatsService;
+import fr.paris.lutece.portal.service.daemon.Daemon;
+
+public class StatsDaemon extends Daemon
+{
+
+    @Override
+    public void run( )
+    {
+        SearchStatsService.sendEmail( );
```
Comment (rzara) [inline src/java/fr/paris/lutece/plugins/searchstats/daemon/StatsDaemon.java]: Some logging would be nice (via `setLastRunLogs(String)`). Maybe at least the number of messages sent ?

File: src/java/fr/paris/lutece/plugins/searchstats/business/QueryRecordDAO.java [hunk 17]
```diff
@@ -186,4 +196,28 @@ private String buildQuery( String strSelect, List<String> clauses )
 
         return strSQL;
     }
+
+    public List<QueryRecord> selectQueryRecordList( Plugin plugin )
```
Comment (rzara) [inline src/java/fr/paris/lutece/plugins/searchstats/business/QueryRecordDAO.java]: How big this list can be ? It seems that you never use the full list anyway. To limit risks of blowing up memory, I think you should pass down restrictions to the search down to here (top n and date criteria).
Maybe extend `RecordFilter` with your needs, and use `selectQueryRecordListByCriteria`

File: src/java/fr/paris/lutece/plugins/searchstats/business/IQueryRecordDAO.java [hunk 18]
```diff
@@ -47,23 +47,33 @@
...
      * Load the list of RecordCount for each date that have at least a record
-     * @param plugin The Plugin using this data access service
+     * 
+     * @param plugin
+     *            The Plugin using this data access service
      * @return The Collection of the RecordCount
      */
     public List<RecordCount> selectQueryRecordDatesList( Plugin plugin, boolean bNoResult );
+
+    //
```
Comment (rzara) [inline src/java/fr/paris/lutece/plugins/searchstats/business/IQueryRecordDAO.java]: Missing javadoc comment

File: src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java [hunk 19]
```diff
@@ -0,0 +1,143 @@
+package fr.paris.lutece.plugins.searchstats.service;
```
Comment (rzara) [inline src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java]: License is missing

File: src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+        return timedRecord;
+    }
+
+    /**
+     *
+     * @param listQueryRecords
+     * @param nLimit
+     * @return The list of the QueryRecords after the last daemon
+     */
+    public List<TermRecord> getMapSearchStatsAfterLastDaemon( List<QueryRecord> listQueryRecords, int nLimit )
```
Comment (rzara) [inline src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java]: What does Map mean in this method name ?

File: src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java [hunk 21]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    /**
+     * @param timeReport
+     *            The time from which the query can be taken into account
+     * @param query
+     * @return Verify if the queriy has been done after the timeReport
+     */
+
+    public static boolean checkAfterDate( TimedRecord timeReport, QueryRecord query )
```
Comment (rzara) [inline src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java]: You are not passing a date but a duration. Maybe find a better name for this method ?

File: src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java [hunk 22]
```diff
@@ -0,0 +1,143 @@
...
+        if ( timeReport.getYear( ) == 0 && timeReport.getMonth( ) == 0 && timeReport.getDay( ) == 0 )
+        {
+            timeReport.setDay( 1 );
+        }
+        return getMapSearchStatsAfterdate( timeReport.getYear( ), timeReport.getMonth( ), timeReport.getDay( ), timeReport.getHour( ), listQueryRecords, nLimit );
+    }
+
+    /**
+     * @param timeReport
+     *            The time from which the query can be taken into account
```
Comment (rzara) [inline src/java/fr/paris/lutece/plugins/searchstats/service/SearchStatsService.java]: It seems to me that this represents a duration (which you then remove from current time to get a new time)

**Comment by jonenst** [comment]
Thanks for the good review @rzara :) @AlexandreClose is overseeing this work.
